File: StarField.java
//
// StarField.java
//

import java.awt.*;

public class StarField {

  // -- Constants --

  /** Number of stars in star field. */
  private static final int STARS = 50;

  /** Minimum star speed. */
  private static final int STAR_MIN = 1;

  /** Maximum star speed. */
  private static final int STAR_MAX = 4;


  // -- Fields --

  /** Dimensions of star field. */
  private int width, height;

  /** Star field Y values. */
  private double[] starY;

  /** Star field velocity values. */
  private double[] starV;


  // -- Constructor --

  /** Constructs a star field. */
  public StarField(int w, int h) {
    width = w; height = h;
    starY = new double[STARS];
    starV = new double[STARS];
    for (int i=0; i<STARS; i++) {
      starY[i] = height * Math.random();
      starV[i] = (STAR_MAX - STAR_MIN + 1) * Math.random() + STAR_MIN;
    }
  }


  // -- StarField API methods --

  /** Draws stars to the given graphics context. */
  public void drawStars(Graphics g) {
    g.setColor(Color.gray);
    for (int i=0; i<STARS; i++) {
      int x = i * width / STARS;
      g.drawLine(x, (int) starY[i], x, (int) (starY[i] - 2 * starV[i]));
    }
  }

  /** Updates star positions. */
  public void moveStars() {
    for (int i=0; i<STARS; i++) {
      starY[i] += starV[i];
      if (starY[i] - 2 * starV[i] > height) {
        starY[i] = 0;
        starV[i] = (STAR_MAX - STAR_MIN + 1) * Math.random() + STAR_MIN;
      }
    }
  }

}

File: js/star-field.js
class StarField {
  constructor(w, h) {
    this.count = 50;   // Number of stars in star field.
    this.minSpeed = 1; // Minimum star speed.
    this.maxSpeed = 4; // Maximum star speed.

    this.width = w;
    this.height = h;
    this.starY = []; // Y values
    this.starV = []; // velocities
    for (var i=0; i<this.count; i++) {
      this.starY[i] = this.height * Math.random();
      this.starV[i] = (this.maxSpeed - this.minSpeed + 1) * Math.random() + this.minSpeed;
    }
  }

  /** Draws stars to the given graphics context. */
  drawStars(ctx) {
    for (var i=0; i<this.count; i++) {
      var x = i * this.width / this.count;
      ctx.beginPath();
      ctx.strokeStyle = 'gray';
      ctx.moveTo(x, this.starY[i]);
      ctx.lineTo(x, this.starY[i] - 2 * this.starV[i]);
      ctx.stroke();
    }
  }

  /** Updates star positions. */
  moveStars() {
    for (var i=0; i<this.count; i++) {
      this.starY[i] += this.starV[i];
      if (this.starY[i] - 2 * this.starV[i] > this.height) {
        this.starY[i] = 0;
        this.starV[i] = (this.maxSpeed - this.minSpeed + 1) * Math.random() + this.minSpeed;
      }
    }
  }
}
